synapse/push/pusherpool.py

```diff
@@ -199,19 +203,45 @@ async def remove_pushers_by_app_id_and_pushkey_not_user(
                 )
                 await self.remove_pusher(p.app_id, p.pushkey, p.user_name)
 
-    async def remove_pushers_by_access_token(
+    async def remove_http_pushers_by_access_tokens(
         self, user_id: str, access_tokens: Iterable[int]
     ) -> None:
-        """Remove the pushers for a given user corresponding to a set of
+        """Remove the HTTP pushers for a given user corresponding to a set of
         access_tokens.
 
         Args:
             user_id: user to remove pushers for
             access_tokens: access token *ids* to remove pushers for
         """
+        # XXX(quenting): This is only needed until the "set_device_id_for_pushers"
+        # background update finishes
         tokens = set(access_tokens)
         for p in await self.store.get_pushers_by_user_id(user_id):
-            if p.access_token in tokens:
+            if p.kind == "http" and p.access_token in tokens:
+                logger.info(
+                    "Removing pusher for app id %s, pushkey %s, user %s",
+                    p.app_id,
+                    p.pushkey,
+                    p.user_name,
+                )
+                await self.remove_pusher(p.app_id, p.pushkey, p.user_name)
+
+    async def remove_http_pushers_by_devices(
+        self, user_id: str, devices: Iterable[str]
+    ) -> None:
+        """Remove the HTTP pushers for a given user corresponding to a set of devices
+
+        Args:
+            user_id: user to remove pushers for
+            devices: device IDs to remove pushers for
+        """
+        device_ids = set(devices)
+        for p in await self.store.get_pushers_by_user_id(user_id):
+            if (
+                p.kind == "http"
+                and p.device_id is not None
+                and p.device_id in device_ids
```

Comment:
`None` can't be in the input so I think this should work fine?

```suggestion
                and p.device_id in device_ids
```